src/osd/OSDMap.cc

```diff
@@ -937,9 +937,18 @@ void OSDMap::set_max_osd(int m)
 int OSDMap::calc_num_osds()
 {
   num_osd = 0;
-  for (int i=0; i<max_osd; i++)
+  num_up_osd = 0;
+  num_in_osd = 0;
+  for (int i=0; i<max_osd; i++) {
     if (osd_state[i] & CEPH_OSD_EXISTS)
-      num_osd++;
+      ++num_osd;
+    if ((osd_state[i] & CEPH_OSD_EXISTS) &&
+	(osd_state[i] & CEPH_OSD_UP))
+      ++num_up_osd;
+    if ((osd_state[i] & CEPH_OSD_EXISTS) &&
+	get_weight(i) != CEPH_OSD_OUT)
+      ++num_in_osd;
```

Comment:
Not terribly important (and probably optimized away) but why not nest these instead of double-testing the exists check?